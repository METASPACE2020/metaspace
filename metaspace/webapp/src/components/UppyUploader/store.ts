// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from 'vue'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State | any>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach((listener) => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
