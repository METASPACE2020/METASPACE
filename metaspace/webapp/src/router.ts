/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter, { RawLocation } from 'vue-router'
import AboutPage from './modules/App/AboutPage.vue'
import DatasetsPage from './modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from './modules/Account'
import { redirectAfterSignIn } from './modules/Account/signInReturnUrl'
import { updateDBParam } from './modules/Filters/url'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Annotations/AnnotationsPage.vue'),
  DatasetSummary: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ './modules/Datasets/summary/DatasetSummary.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ './modules/MetadataEditor/MetadataEditPage.vue'),
  ImageAlignmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ './modules/ImageAlignment/ImageAlignmentPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ './modules/MetadataEditor/UploadPage.vue'),
  DesignPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DesignPage" */ './design/DesignPage.vue'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/HelpPage.vue'),
  EditUserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/GroupProfile/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/GroupsListPage.vue'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/TermsPage.vue'),
  PublicationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PublicationsPage.vue'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/GroupProfile/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Project/ViewProjectPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(about)?', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
    { path: '/annotations', component: asyncPages.AnnotationsPage },
    {
      path: '/datasets',
      component: DatasetsPage,
      children: [
        { path: '', component: asyncPages.DatasetTable },
        { path: 'summary', component: asyncPages.DatasetSummary },
      ],
    },
    { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
    { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.ImageAlignmentPage },
    { path: '/upload', component: asyncPages.UploadPage },
    { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },
    { path: '/user/me', component: asyncPages.EditUserPage },

    { path: '/admin/health', component: asyncPages.SystemHealthPage },
    { path: '/admin/groups', component: asyncPages.GroupsListPage },

    { path: '/account/sign-in', component: DialogPage, props: { dialog: 'signIn' } },
    { path: '/account/sign-in-success', redirect: redirectAfterSignIn },
    { path: '/account/create-account', component: DialogPage, props: { dialog: 'createAccount' } },
    { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
    { path: '/account/reset-password', component: ResetPasswordPage },

    { path: '/group/create', component: asyncPages.CreateGroupPage },
    { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

    { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
    { // Legacy URL sent in "request access" emails up until Feb 2019
      path: '/project/:projectIdOrSlug/manage',
      redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } } as RawLocation,
    },
    { path: '/projects', component: asyncPages.ProjectsListPage },

    { path: '/terms', component: asyncPages.TermsPage, meta: { footer: true } },
    { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
    { path: '/publications', component: asyncPages.PublicationsPage, meta: { footer: true } },

    { path: '/design', component: asyncPages.DesignPage, meta: { footer: true, flex: true } },

    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
