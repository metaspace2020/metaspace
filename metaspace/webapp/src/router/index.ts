/* eslint-disable vue/max-len */
import { createRouter, createWebHistory } from 'vue-router'
import { Component } from 'vue'
import AboutPage from '../modules/App/AboutPage'
import DatasetsPage from '../modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from '../modules/Account'
import { redirectAfterSignIn } from '../modules/Account/signInReturnUrl'
import { updateDBParam } from '../modules/Filters/url'
import NotFound from '../modules/App/NotFoundPage.vue'

const asyncPagesFreelyTyped = {
  AnnotationsPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ '../modules/Annotations/AnnotationsPage.vue'
    ),
  DatasetSummary: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ '../modules/Datasets/summary/DatasetSummary.vue'
    ),
  MetadataEditPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ '../modules/MetadataEditor/MetadataEditPage.vue'
    ),
  ImageAlignmentPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ '../modules/ImageAlignment/ImageAlignmentPage.vue'
    ),
  UploadPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ '../modules/MetadataEditor/UploadPage.vue'),
  DatasetOverviewPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ '../modules/Datasets/overview/DatasetOverviewPage'
    ),
  DatasetComparisonPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "DatasetComparisonPage" */ '../modules/Datasets/comparison/DatasetComparisonPage'
    ),
  DatasetBrowserPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "DatasetBrowserPage" */ '../modules/Datasets/imzml/DatasetBrowserPage'
    ),
  DatasetEnrichmentPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "DatasetEnrichmentPage" */ '../modules/Datasets/enrichment/DatasetEnrichmentPage'
    ),
  SpottingProjectPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "SpottingProjectPage" */ '../modules/SpottingProject/DashboardPage'
    ),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/HelpPage.vue'),
  EditUserPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Group/CreateGroupPage.vue'),
  ProjectsListPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Group/GroupsListPage'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/TermsPage.vue'),
  PublicationsPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/PublicationsPage'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ '../modules/Group/ViewGroupPage.vue'),
  ViewProjectPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ '../modules/Project/ViewProjectPage.vue'),

  // Separate bundle for design docs
  DesignTOC: () => import(/* webpackChunkName: "DesignBundle" */ '../design/TOCPage.vue'),
  DesignStyleGuide: () => import(/* webpackChunkName: "DesignBundle" */ '../design/StyleGuidePage.vue'),
  DesignIcons: () => import(/* webpackChunkName: "DesignBundle" */ '../design/IconsPage.vue'),
  DesignComponents: () => import(/* webpackChunkName: "DesignBundle" */ '../design/ComponentsPage.vue'),
  DesignForms: () => import(/* webpackChunkName: "DesignBundle" */ '../design/FormsPage.vue'),
  ContactPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ContactPage" */ '../modules/Contact/Contact'),
}

const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, Component>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

export const routes: any = [
  { path: '/', name: 'home', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
  { path: '/about', name: 'about', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
  { path: '/annotations', name: 'annotations', component: asyncPages.AnnotationsPage },
  { path: '/contact', name: 'contact', component: asyncPages.ContactPage },
  {
    path: '/datasets',
    name: 'dataset-list',
    component: DatasetsPage,
    children: [
      { path: '', name: 'datasets', component: asyncPages.DatasetTable },
      { path: 'summary', name: 'summary', component: asyncPages.DatasetSummary },
    ],
  },
  {
    path: '/datasets/:dataset_id/comparison',
    name: 'datasets-comparison',
    component: asyncPages.DatasetComparisonPage,
  },
  { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
  {
    path: '/datasets/:dataset_id/add-optical-image',
    name: 'add-optical-image',
    component: asyncPages.ImageAlignmentPage,
  },
  { path: '/dataset/:dataset_id', name: 'dataset-overview', component: asyncPages.DatasetOverviewPage },
  { path: '/dataset/:dataset_id/annotations', name: 'dataset-annotations', component: asyncPages.AnnotationsPage },
  { path: '/dataset/:dataset_id/browser', name: 'dataset-browser', component: asyncPages.DatasetBrowserPage },
  { path: '/dataset/:dataset_id/enrichment', name: 'dataset-enrichment', component: asyncPages.DatasetEnrichmentPage },
  { path: '/upload', name: 'upload', component: asyncPages.UploadPage },
  { path: '/help', name: 'help', component: asyncPages.HelpPage, meta: { footer: true } },
  { path: '/user/me', name: 'profile', component: asyncPages.EditUserPage },

  { path: '/admin/health', name: 'admin', component: asyncPages.SystemHealthPage },

  { path: '/account/sign-in', name: 'sign-in', component: DialogPage, props: { dialog: 'signIn' } },
  { path: '/account/sign-in-success', name: 'sign-in-success', redirect: redirectAfterSignIn },
  { path: '/account/create-account', name: 'create-acc', component: DialogPage, props: { dialog: 'createAccount' } },
  { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
  { path: '/account/reset-password', component: ResetPasswordPage },

  { path: '/groups', name: 'group-list', component: asyncPages.GroupsListPage },
  { path: '/group/create', name: 'group-create', component: asyncPages.CreateGroupPage },
  { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

  { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
  {
    // Legacy URL sent in "request access" emails up until Feb 2019
    path: '/project/:projectIdOrSlug/manage',
    redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } },
  },
  { path: '/projects', name: 'project-list', component: asyncPages.ProjectsListPage },

  { path: '/terms', name: 'terms', component: asyncPages.TermsPage, meta: { footer: true } },
  { path: '/privacy', name: 'privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
  { path: '/publications', name: 'publication-list', component: asyncPages.PublicationsPage, meta: { footer: true } },

  { path: '/design', name: 'design', component: asyncPages.DesignTOC, meta: { footer: true, flex: true } },
  { path: '/design/styleguide', component: asyncPages.DesignStyleGuide, meta: { footer: true, flex: true } },
  { path: '/design/icons', component: asyncPages.DesignIcons, meta: { footer: true, flex: true } },
  { path: '/design/components', component: asyncPages.DesignComponents, meta: { footer: true, flex: true } },
  { path: '/design/forms', component: asyncPages.DesignForms, meta: { footer: true, flex: true } },

  { path: '/detectability', name: 'detectability', component: asyncPages.SpottingProjectPage },

  { path: '/:pathMatch(.*)*', name: 'not-found', component: NotFound, meta: { footer: true, flex: true } },
]

const router = createRouter({
  // @ts-ignore
  history: createWebHistory(import.meta.env.BASE_URL),
  routes,
  scrollBehavior(to, from, savedPosition) {
    // return desired position
    if (savedPosition) {
      return savedPosition
    }
    return { top: 0 }
  },
})

const pageLoadedAt = Date.now()
router.beforeEach((to, from, next) => {
  // If user has had the window open for a long time, it's likely they're using old code. This can cause issues
  // such as incorrect API usage, or causing error reports for bugs that have already been fixed.
  // Force the browser to reload the page, preferring to do so during page transitions to minimize impact.
  const daysSincePageLoad = (Date.now() - pageLoadedAt) / 1000 / 60 / 60 / 24
  if ((to.path !== from.path && daysSincePageLoad > 1.75) || daysSincePageLoad > 7) {
    window.location.assign(to.fullPath)
    next(false)
  } else {
    next()
  }
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
