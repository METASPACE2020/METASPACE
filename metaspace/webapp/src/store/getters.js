import {mzFilterPrecision} from '../util';
import {decodeParams, decodeSettings, getLevel} from '../modules/Filters';
import config from '../config';

/** For filters where empty string is a valid client-side value that must be converted to empty string for the API */
const noneToEmptyString = s => s === 'none' ? '' : s;

export default {
  filterLevel(state) {
    return getLevel(state.route.path);
  },

  filter(state) {
    return decodeParams(state.route, state.filterLists);
  },

  settings(state) {
    return decodeSettings(state.route);
  },

  ftsQuery(state, getters) {
    return getters.filter.simpleQuery;
  },

  gqlAnnotationFilter(state, getters) {
    const filter = getters.filter;
    const colocalizationAlgo = getters.settings.annotationView.colocalizationAlgo;

    const f = {
      database: filter.database,
      compoundQuery: filter.compoundName,
      chemMod: noneToEmptyString(filter.chemMod),
      neutralLoss: noneToEmptyString(filter.neutralLoss),
      adduct: filter.adduct,
      fdrLevel: filter.fdrLevel,
      colocalizedWith: filter.colocalizedWith,
      // Only include colocalizationAlgo if there is another filter that uses it. Otherwise the annotations list
      // refreshes unnecessarily when changing algorithm.
      colocalizationAlgo: filter.colocalizedWith || filter.colocalizationSamples ? colocalizationAlgo : null,
      colocalizationSamples: filter.colocalizationSamples,
      offSample: filter.offSample == null ? undefined : !!filter.offSample,
    };

    if (!config.features.neutral_losses && !config.features.advanced_ds_config) {
      f.hasNeutralLoss = false;
    }

    if (!config.features.chem_mods && !config.features.advanced_ds_config) {
      f.hasChemMod = false;
    }

    if (!config.features.all_adducts) {
      f.hasHiddenAdduct = false;
    }

    if (filter.minMSM)
      f.msmScoreFilter = {min: filter.minMSM, max: 1.0};

    if (filter.mz) {
      const mz = parseFloat(filter.mz),
            deltamz = parseFloat(mzFilterPrecision(mz));
      f.mzFilter = {
        min: mz - deltamz,
        max: mz + deltamz
      };
    }

    return f;
  },

  gqlDatasetFilter(state, getters) {
    const filter = getters.filter;
    const {group, project, submitter, datasetIds, polarity,
           organism, organismPart, condition, growthConditions,
           ionisationSource, analyzerType, maldiMatrix, metadataType,
           compoundName} = filter;
    const level = getters.filterLevel;
    const hasAnnotationMatching = level === 'dataset' && compoundName ? { compoundQuery: compoundName } : undefined;
    return {
      group: group,
      project: project,
      submitter: submitter,

      // temporary workaround because of array-related bugs in apollo-client
      ids: datasetIds ? datasetIds.join("|") : null,

      organism,
      organismPart,
      condition,
      growthConditions,
      ionisationSource,
      maldiMatrix,
      analyzerType,
      polarity: polarity ? polarity.toUpperCase() : null,
      metadataType,
      hasAnnotationMatching,
    }
  },

  gqlColocalizationFilter(state, getters) {
    const {datasetIds, colocalizedWith, database, fdrLevel} = getters.filter;
    const colocalizationAlgo = getters.settings.annotationView.colocalizationAlgo;
    if (datasetIds && !datasetIds.includes('|') && colocalizedWith != null && database != null && fdrLevel != null) {
      return {colocalizedWith, colocalizationAlgo, database, fdrLevel};
    } else {
      return null;
    }
  },
}
