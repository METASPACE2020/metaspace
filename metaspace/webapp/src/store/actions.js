import { fetchOptionListsQuery } from '../api/metadata'
import { decodeParams } from '../modules/Filters'
import { computed } from 'vue'
import store from '../store/index'

export default {
  async initFilterLists(context, apolloClient) {
    if (context.state.filterListsLoading || context.state.filterLists != null) return
    await context.commit('setFilterListsLoading')

    const response = await apolloClient.query({
      query: fetchOptionListsQuery,
      fetchPolicy: 'cache-first',
    })

    // set annotationIds default according to values passed from snapshot to the store state
    context.commit('setFilterLists', {
      ...response.data,
      annotationIds: computed(() => store.state.snapshotAnnotationIds),
    })

    // Refresh the current filter so that computed defaults that depend on `filterLists` are applied
    const filter = decodeParams(context.state.route, context.state.filterLists)

    context.commit('updateFilter', filter)
  },
}
