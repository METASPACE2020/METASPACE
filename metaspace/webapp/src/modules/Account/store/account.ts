import { Module } from 'vuex';
import { DialogType } from '../dialogs';
import router from '../../../router';

export interface AccountState {
  dialog: DialogType | null;
  showDialogAsPage: boolean;
}

const account = {
  namespaced: true,
  state: {
    dialog: null,
    showDialogAsPage: false
  },
  mutations: {
    showDialog(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = false;
    },
    showDialogAsPage(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = true;
    },
    hideDialog(state: AccountState, dialog: DialogType) {
      if (state.dialog === dialog) {
        state.dialog = null;
        if (state.showDialogAsPage) {
          router.push('/');
        }
      }
    }
  }
};

export default account as Module<AccountState, any>;
