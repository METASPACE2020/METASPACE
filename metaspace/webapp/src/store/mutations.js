import {isEqual, pull, without} from 'lodash-es';
import router from '../router';
import compare from '../lib/compare';

import {
  decodeParams,
  encodeParams, encodeSections, encodeSortOrder, FILTER_SPECIFICATIONS,
  stripFilteringParams,
} from '../modules/Filters';
import {getFilterInitialValue} from '../modules/Filters';


function updatedLocation(state, filter) {
  let query = encodeParams(filter, state.route.path, state.filterLists);

  state.lastUsedFilters[state.route.path] = {
    filter,
    query,
    order: state.orderedActiveFilters
  };

  return {
    query: Object.assign(query, stripFilteringParams(state.route.query))
  };
}

function replaceURL(state, filter) {
  router.replace(updatedLocation(state, filter));
}

function pushURL(state, filter) {
  router.push(updatedLocation(state, filter));
}

function sortFilterKeys(keys) {
  keys.sort((a, b) => compare(FILTER_SPECIFICATIONS[a].sortOrder || 100, FILTER_SPECIFICATIONS[b].sortOrder || 100));
}

export default {
  updateFilter (state, filter) {
    let active = [];

    // drop unset filters
    for (var i = 0; i < state.orderedActiveFilters.length; i++) {
      let key = state.orderedActiveFilters[i];
      if (filter[key] !== undefined)
        active.push(key);
    }

    // append newly added filters to the end
    for (var key in filter)
      if (filter[key] !== undefined &&
          active.indexOf(key) == -1)
        active.push(key);

    // sort
    sortFilterKeys(active);

    const changedFilterSet = !isEqual(state.orderedActiveFilters, active);

    state.orderedActiveFilters = active;
    if (changedFilterSet)
      pushURL(state, filter);
    else
      replaceURL(state, filter);
  },

  addFilter (state, name) {
    const initialValue = getFilterInitialValue(name, state.filterLists);

    // FIXME: is there any way to access getters here?
    let filter = Object.assign(decodeParams(state.route, state.filterLists),
                               {[name]: initialValue});

    state.orderedActiveFilters.push(name);
    pushURL(state, filter);
  },

  updateFilterOnNavigate(state, to) {
    const newActiveFilters = Object.keys(decodeParams(to, state.filterLists));

    const removed = without(state.orderedActiveFilters, ...newActiveFilters);
    const added = without(newActiveFilters, ...state.orderedActiveFilters);

    pull(state.orderedActiveFilters, ...removed);
    state.orderedActiveFilters.push(...added);
    sortFilterKeys(state.orderedActiveFilters);
  },

  setFilterListsLoading(state) {
    state.filterListsLoading = true;
  },

  setFilterLists(state, filterLists) {
    state.filterLists = filterLists;
    state.filterListsLoading = false;
  },

  setAnnotation(state, annotation) {
    state.annotation = annotation;
  },

  updateAnnotationTableStatus(state, isLoading) {
    state.tableIsLoading = isLoading;
  },

  startTour(state, tourDescription) {
    state.currentTour = tourDescription;
  },

  endTour(state) {
    state.currentTour = null;
  },

  updateAnnotationViewSections(state, activeSections) {
    let query = Object.assign({}, state.route.query, {
      sections: encodeSections(activeSections)
    });
    router.replace({query});
  },

  setColormap(state, colormap) {
    let query = Object.assign({}, state.route.query, {
      cmap: colormap
    });
    router.replace({query});
  },

  setCurrentPage(state, page) {
    let query = Object.assign({}, state.route.query, {page});
    router.replace({query});
  },

  setSortOrder(state, sortOrder) {
    let query = Object.assign({}, state.route.query, {
      sort: encodeSortOrder(sortOrder)
    });
    router.replace({query});
  },

  setDatasetTab(state, tab) {
    const path = {
      'list': '/datasets',
      'summary': '/datasets/summary'
    }[tab.toLowerCase()];
    router.replace({
      path,
      query: state.route.query,
    });
  }
};
