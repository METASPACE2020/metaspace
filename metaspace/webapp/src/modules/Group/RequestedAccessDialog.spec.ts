import { mount } from '@vue/test-utils'
import { defineComponent, nextTick, h } from 'vue'
import { RequestedAccessDialog } from './RequestedAccessDialog'
import router from '../../router'
import store from '../../store'

describe('RequestedAccessDialog', () => {
  const propsData = {
    group: {
      id: '7871c940-8198-11eb-8245-9b435184cf72',
      name: 'TEST_GROUP',
      shortName: 'TG',
      urlSlug: 'tgroup',
      currentUserRole: 'PENDING',
      members: [
        {
          role: 'GROUP_ADMIN',
          numDatasets: 2,
          user: {
            id: '039801c8-919e-11eb-908e-3b2b8e672707',
            name: 'John Doe',
            email: 'jdoe@mail.com',
          },
        },
        {
          role: 'GROUP_ADMIN',
          numDatasets: 2,
          user: {
            id: '039801c8-919e-11eb-908e-3b2b8e672701',
            name: 'John Doe 2',
            email: 'jdoe2@mail.com',
          },
        },
        {
          role: 'GROUP_ADMIN',
          numDatasets: 2,
          user: {
            id: '039801c8-919e-11eb-908e-3b2b8e672702',
            name: 'John Doe 3',
            email: 'jdoe3@mail.com',
          },
        },
      ],
    },
    visible: true,
  }

  const testHarness = defineComponent({
    components: {
      RequestedAccessDialog,
    },
    props: ['group', 'visible', 'dsSubmission'],
    setup(props) {
      return () => h(RequestedAccessDialog, { ...props })
    },
  })

  it('it should match snapshot', async () => {
    const wrapper = mount(testHarness, {
      global: {
        plugins: [store, router],
      },
      props: propsData,
    })
    await nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('it should match dataset processing snapshot', async () => {
    const wrapper = mount(testHarness, {
      global: {
        plugins: [store, router],
      },
      props: {
        ...propsData,
        dsSubmission: true,
      },
    })
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })
})
