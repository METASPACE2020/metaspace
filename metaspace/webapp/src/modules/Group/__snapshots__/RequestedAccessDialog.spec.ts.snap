// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RequestedAccessDialog > it should match dataset processing snapshot 1`] = `"<div class=\\"mock-el-dialog requested-access-dialog\\" model-value=\\"true\\" append-to-body=\\"true\\" title=\\"TEST_GROUP Access Request\\" lockscroll=\\"false\\">Your dataset was sent to processing. Please contact<span class=\\"ml-1\\"><a href=\\"mailto:jdoe@mail.com\\">John Doe</a>,</span><span class=\\"ml-1\\"><a href=\\"mailto:jdoe2@mail.com\\">John Doe 2</a> or</span><span class=\\"ml-1\\"><a href=\\"mailto:jdoe3@mail.com\\">John Doe 3</a> </span>to approve your request within METASPACE, so you can visualize the group datasets!</div>"`;

exports[`RequestedAccessDialog > it should match snapshot 1`] = `"<div class=\\"mock-el-dialog requested-access-dialog\\" model-value=\\"true\\" append-to-body=\\"true\\" title=\\"TEST_GROUP Access Request\\" lockscroll=\\"false\\">Your request was sent. Please contact<span class=\\"ml-1\\"><a href=\\"mailto:jdoe@mail.com\\">John Doe</a>,</span><span class=\\"ml-1\\"><a href=\\"mailto:jdoe2@mail.com\\">John Doe 2</a> or</span><span class=\\"ml-1\\"><a href=\\"mailto:jdoe3@mail.com\\">John Doe 3</a> </span>to approve your request within METASPACE, so you can visualize the group datasets!</div>"`;
